tensorflow/c/experimental/profiler/profiler.cc

```diff
@@ -0,0 +1,225 @@
+  // * At the minimum we want to disallow '/' and ':' since
+  //   these characters are used in device spec, for e.g.
+  //   /job:foo/replica:12/device:GPU:1.
+  // * Underscores seem useful, for e.g. XLA_GPU uses underscores.
+  // * Allowing lowercase might get confusing. For example, say someone
+  //   registers a new type called "Gpu". It might be confusing for users that
+  //   "Gpu" is not the same device type as "GPU".
+  //   Note that lowercase "cpu" and "gpu" are currently supported only for
+  //   legacy reasons:
+  //   https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/python/framework/device_spec.py;l=46;drc=d3a378f9665d8eee827c74cb9ecbee81e4c288dd
+  static const LazyRE2 kTfDeviceTypeRegEx = {"[A-Z][A-Z_]*"};
+  bool matches = RE2::FullMatch(type, *kTfDeviceTypeRegEx);
+  if (!matches) {
+    return Status(
+        error::FAILED_PRECONDITION,
+        tensorflow::strings::StrCat("Device name/type '", type, "' must match ",
+                                    kTfDeviceTypeRegEx->pattern(), "."));
+  }
+  return Status::OK();
+}
+
+Status ValidateTPProfilerRegistrationParams(
+    const TF_ProfilerRegistrationParams& params) {
+  VALIDATE_STRUCT_SIZE(TF_ProfilerRegistrationParams, params,
+                       TF_PROFILER_REGISTRATION_PARAMS_STRUCT_SIZE);
+  VALIDATE_MEMBER(TF_ProfilerRegistrationParams, params, destroy_profiler);
+  VALIDATE_MEMBER(TF_ProfilerRegistrationParams, params, destroy_profiler_fns);
+  return Status::OK();
+}
+
+Status ValidateTPProfiler(const TP_Profiler& profiler) {
+  VALIDATE_STRUCT_SIZE(TP_Profiler, profiler, TP_PROFILER_STRUCT_SIZE);
+  VALIDATE_MEMBER(TP_Profiler, profiler, type);
+  TF_RETURN_IF_ERROR(ValidateDeviceType(profiler.type));
+  return Status::OK();
+}
+
+Status ValidateTPProfilerFns(const TP_ProfilerFns& profiler_fns) {
+  VALIDATE_STRUCT_SIZE(TP_ProfilerFns, profiler_fns,
+                       TF_PROFILER_FNS_STRUCT_SIZE);
+  VALIDATE_MEMBER(TP_ProfilerFns, profiler_fns, start);
+  VALIDATE_MEMBER(TP_ProfilerFns, profiler_fns, stop);
+  VALIDATE_MEMBER(TP_ProfilerFns, profiler_fns, collect_data_xspace);
+  return Status::OK();
+}
+
+class PluggableProfiler : public tensorflow::profiler::ProfilerInterface {
+ public:
+  // The caller must have validated profiler_fns and profiler.
+  static std::unique_ptr<tensorflow::profiler::ProfilerInterface>
+  CreatePluggableProfiler(const ProfileOptions& options, TP_Profiler profiler,
+                          TP_ProfilerFns profiler_fns) {
+    if (options.device_tracer_level() == 0) {
+      return nullptr;
+    }
+    if (options.device_type() != ProfileOptions::UNSPECIFIED &&
+        options.device_type() != ProfileOptions::PLUGGABLE_DEVICE) {
+      return nullptr;
+    }
+    return absl::WrapUnique(new PluggableProfiler(profiler_fns, profiler));
+  }
+
+  Status Start() {
+    std::unique_ptr<TF_Status, TFStatusDeleter> c_status(TF_NewStatus());
```

Comment:
The pattern I see elsewhere in the code base is:

std::unique_ptr<TF_Status, decltype(&TF_DeleteStatus)> status(
      TF_NewStatus(), TF_DeleteStatus);

I would also name this variable status instead of c_status, the prefix does not add much value in this case.

Reply:
done